clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -740,4 +728,148 @@ public double measure(MetricConfig config, long now) {
+                                public void onFailure(RuntimeException e) {
+
+                                }
+                            });
+                        } else if (heartbeat.sessionTimeoutExpired(now)) {
+                            // the session timeout has expired without seeing a successful heartbeat, so we should
+                            // probably make sure the coordinator is still healthy.
+                            coordinatorDead();
+                        } else if (heartbeat.pollTimeoutExpired(now)) {
+                            // the poll timeout has expired, which means that the foreground thread has stalled
+                            // in between calls to poll(), so we explicitly leave the group.
+                            maybeLeaveGroup();
+                        } else if (!heartbeat.shouldHeartbeat(now)) {
+                            awaitMs(heartbeat.timeToNextHeartbeat(now));
+                        } else {
+                            heartbeat.sentHeartbeat(now);
+                            RequestFuture<Void> future = sendHeartbeatRequest();
+                            future.addListener(new RequestFutureListener<Void>() {
+                                @Override
+                                public void onSuccess(Void value) {
+                                    heartbeat.receiveHeartbeat(time.milliseconds());
+                                }
+
+                                @Override
+                                public void onFailure(RuntimeException e) {
+                                    if (e instanceof RebalanceInProgressException) {
+                                        // it is valid to continue heartbeating while the group is rebalancing. This
+                                        // ensures that the coordinator keeps the member in the group for as long
+                                        // as the duration of the rebalance timeout. If we stop sending heartbeats,
+                                        // however, then the session timeout may expire before we can rejoin.
+                                        heartbeat.receiveHeartbeat(time.milliseconds());
+                                    } else {
+                                        heartbeat.failHeartbeat();
+
+                                        // wake up the thread if it's sleeping to reschedule the heartbeat
+                                        synchronized (AbstractCoordinator.this) {
+                                            AbstractCoordinator.this.notify();
+                                        }
+                                    }
+                                }
+                            });
+                        }
+                    }
+                }
+            } catch (RuntimeException e) {
+                log.error("Heartbeat thread for group {} failed due to unexpected error" , groupId, e);
+                this.failed.set(e);
```

Comment:
So this will set the exception, which will then be rethrown in the user thread. But then what? It seems like there's no way to reset things after that? Is the assumption that the consumer should be considered in a bad state and you have to allocated a new one? How do you tell that from the exception that's thrown. Or could we kill the thread and allow it to be reallocated before throwing the exception in `pollHeartbeat`?